### ISSUE TITLE
Interactive playground for einops

### ISSUE THREAD

**Issue by arogozhnikov** [arogozhnikov/einops#340]
I want to use interactive notebooks as an intro to einops, currently there are two versions: on github, and converted version in docs.

Only two notebooks work in jupyterlite (basic operations, and pack/unpack), but that's already enough for an intro.

Trying to figure out a simple setup to 

- keep notebooks in github
- do not conflict with gh-pages branch
- auto-update to latest pushed version on github 
- provide utils / npy together with tutorial

**Comment by zachlipp** [comment]
Hey @arogozhnikov! I would love to help contribute to this if you're open to it

**Comment by arogozhnikov** [comment]
@zachlipp yes, some help would be great. 

FYI I expect this to be challenging: I didn't see jupyterlite and mkdocs being co-deployed.

Prefect solution is to have jupyterlite and notebooks being stored in different repos: after loading in browser jupyterlite would download fresh notebooks from einops repository. This would be very reliable and simple to maintain.

**Comment by arogozhnikov** [comment]
I think other projects would be interested in just relying on some github-hosted server if they could just provide a folder with notebooks.

**Comment by arogozhnikov** [comment]
@zachlipp  yes, this would be great setup if it works

**Comment by zachlipp** [comment]
@arogozhnikov I'm making some progress here! One thing to think about: When should we overwrite the files the user has downloaded? I don't think overwriting the notebooks each time the user refreshes the page is a good user experience. 

Maybe we get the git hash of the latest commit when we download the files, store the hash, then prompt the user to update if we see there's a new hash on the main branch?

**Comment by arogozhnikov** [comment]
Wow, great to hear that.

I normally don't expect js playgrounds to store my results, I think this would work.

> Maybe we get the git hash of the latest commit when we download the files, store the hash, then prompt the user to update if we see there's a new hash on the main branch?

Yes, that sounds a a good tradeoff. 
NB: if user reloads page, and you keep user's previous result, it worth saying this to avoid cases when users complain about broken example notebooks (broken because user previously played with it, but didn't realize that's their changes).

**Comment by zachlipp** [comment]
I've spent a few more hours on this, and you're right, it's pretty complicated. I am writing a lot of javascript and I am far from a javascript expert. One thing to flag: Most of the existing tutorials cannot work because pytorch isn't yet supported in pyodide. This is an area of active discussion here https://github.com/pyodide/pyodide/issues/1625

Once I have a hosted environment that can execute some of the examples (I'm close!), I'll let you know and we can decide on appropriate steps to take from there

**Comment by arogozhnikov** [comment]
> pytorch isn't yet supported in pyodide.

yes, we should only plan for two notebooks with numpy (basic into and pack/unpack). Completely fine for playground, and I don't expect DL frameworks to target WASM anytime soon

**Comment by zachlipp** [comment]
I totally agree :)

**Comment by arogozhnikov** [comment]
heyhey @zachlipp how is the progress? 

i've tried this link: zachlipp.github.io/einops-documentation/lab/index.html

that's what appears in chrome console:
```
Error fetching file 2-einops-for-deep-learning.ipynb: DOMException: Failed to execute 'transaction' on 'IDBDatabase': One of the specified object stores was not found.
```

**Comment by zachlipp** [comment]
Here's what I see as the next steps:
- Finishing the JS to properly download the files and store them in `IndexedDB`
- Setting up some sort of limits on when the files get redownloaded. I don't have a great answer here.
- Including the subdirectories in the download
- Excluding files with torch dependencies in the download
- Getting the notebooks we download ready to work with einops out of the gate. That could mean:
  - Pre-installing einops with emscripten forge (instructions [here](https://jupyterlite.readthedocs.io/en/latest/howto/xeus-python/preinstalled_packages.html))
  - Do something clever to run `micropip.install("einops")` before/on kernel load
- Replace the hacky JS + awk pipeline I have now to an actual pyodide module

I am trying to complete these in roughly this order; I definitely welcome help from folks with more frontend experience

**Comment by aldinash** [comment]
Hey @arogozhnikov! I would like to work on this if you're still open to it.

**Comment by arogozhnikov** [comment]
sure, @aldinash, this is still very relevant and interesting thing to implement

**Comment by aldinash** [comment]
Great!
I read through the thread and still have questions about your view on this:
- would it be acceptable to trigger a JupyterLite redeploy whenever the notebooks repo changes?  so fresh notebooks are fetched only on deployment and not every time user visits the website
- do you want the interface be different from the [default](https://jupyterlite.github.io/demo/lab/index.html) one? I guess we can just try to remove unnecessary functionality like uploading notebooks?
- since the notebooks serve as a playground, i think it should be appropriate to reset them on refresh?

**Comment by arogozhnikov** [comment]
> would it be acceptable to trigger a JupyterLite redeploy whenever the notebooks repo changes? 

We should avoid that, I'd really like if I could just pass a link to a folder with notebooks and they would be fetched on page load from github. 
Nbviewer handles this very well: https://nbviewer.org/github/arogozhnikov/einops/tree/main/docs/
(yes it's not interactive)

> do you want the interface be different from the [default](https://jupyterlite.github.io/demo/lab/index.html) one? 

default is totally fine

> since the notebooks serve as a playground, i think it should be appropriate to reset them on refresh? 

yes, that's fine with me
